docs/networking.adoc

```diff
@@ -0,0 +1,766 @@
+fe80::/10                   fe80::aa40:25ff:fe04:35f    U       3   52411 cxgbe1
+fe80::/10                   fe80::aa40:25ff:fe04:357    U       3   59075 cxgbe0
+----
+
+There's a lot to unpack here.  There are two important groups of routes here:
+
+* `fd00:1122:3344::101/64` through `fd00:1122:3344::107/64`: routes for the underlay network (12 routes)
+* `fd00:1122:3344::1/64` through `fd00:1122:3344::3/64`: routes for the internal DNS servers (6 routes)
+* `fdb0:...`: routes for the bootstrap network (21 routes)
+
+If we sort the underlay routes by destination address, we notice that each prefix has two routes: one through cxgbe0 and one through cxgbe1:
+
+[source,console]
+----
+BRM44220005 # netstat -rn -f inet6 | grep ^fd00:1122 | sort
+fd00:1122:3344:101::/64     fe80::aa40:25ff:fe05:40c    UG      2 14094545 cxgbe1
+fd00:1122:3344:101::/64     fe80::aa40:25ff:fe05:c      UG      2  643952 cxgbe0
+fd00:1122:3344:102::/64     fe80::aa40:25ff:fe05:40c    UG      2   12352 cxgbe1
+fd00:1122:3344:102::/64     fe80::aa40:25ff:fe05:c      UG      2 6354653 cxgbe0
+fd00:1122:3344:103::/64     fe80::aa40:25ff:fe05:40c    UG      2   32969 cxgbe1
+fd00:1122:3344:103::/64     fe80::aa40:25ff:fe05:c      UG      2 19512229 cxgbe0
+fd00:1122:3344:104::/64     fd00:1122:3344:104::1       U      22 229164829 underlay0
+fd00:1122:3344:105::/64     fe80::aa40:25ff:fe05:40c    UG      2      47 cxgbe1
+fd00:1122:3344:105::/64     fe80::aa40:25ff:fe05:c      UG      2  134024 cxgbe0
+fd00:1122:3344:106::/64     fe80::aa40:25ff:fe05:40c    UG      2 18678270 cxgbe1
+fd00:1122:3344:106::/64     fe80::aa40:25ff:fe05:c      UG      2 3848371 cxgbe0
+fd00:1122:3344:107::/64     fe80::aa40:25ff:fe05:40c    UG      2 75927564 cxgbe1
+fd00:1122:3344:1::/64       fe80::aa40:25ff:fe05:40c    UG      2      51 cxgbe1
+fd00:1122:3344:1::/64       fe80::aa40:25ff:fe05:c      UG      2    2437 cxgbe0
+fd00:1122:3344:2::/64       fe80::aa40:25ff:fe05:40c    UG      2      38 cxgbe1
+fd00:1122:3344:2::/64       fe80::aa40:25ff:fe05:c      UG      2    2437 cxgbe0
+fd00:1122:3344:3::/64       fe80::aa40:25ff:fe05:40c    UG      2   13358 cxgbe1
+fd00:1122:3344:3::/64       fe80::aa40:25ff:fe05:c      UG      2    2437 cxgbe0
+----
+
+Recall that cxgbe0 and cxgbe1 are connected to separate switches in the rack.  So we're seeing the prefixes for the other sleds in this deployment.  We have two routes to reach each sled: one through each switch.  The gateway is the link-local address _of each switch_ on the corresponding link.  One notable exception: the route for this same sled (`fd00:1122:3344:104::/64`) points to `underlay0`, the GZ's VNIC on the sled's underlay network.  In this way, traffic leaving the GZ (whether it originated in this GZ or arrived from one of the switches) is directed to the sled's underlay network etherstub and from there to the right zone VNIC.
+
+(Questions: Why does 107 only have one route?  What are the `fd00:99::` routes?)
```

Comment:
I think the fd00:99 is a holdover from how we were reaching boundary services for a bit. This is being overhauled per [RFD 404 Boundary Tunnel Routing](https://rfd.shared.oxide.computer/rfd/0404) so it may have something to do with what's provisioned or otherwise. I'm not sure. @rcgoodfellow may have a better idea here.

Reply:
While 404 is in flux, I can take a stab at what the current state looks like:

`fd00:99::1/64` is the boundary services address. The small twist there is that there's no single "boundary services" node present on the network but rather, OPTE and the P4 on the switch essentially have a contract in how external traffic gets transited.

If you query the switch, it should appear as a `loopback` address, i.e. it's not associated with any backplane or front panel links but it is an address the switch is aware of:
```
$ pfexec zlogin oxz_switch /opt/oxide/dendrite/bin/swadm addr ls
Link      IPv4         IPv6
qsfp0/0   172.17.16.2  
loopback  fd00:99::1
```

If we look at an OPTE port for an instance or service with external connectivity, we can see all the layers that process incoming/outgoing packets:
```
$ pfexec opteadm list-layers -p opte7
NAME         RULES IN   RULES OUT  DEF IN   DEF OUT  FLOWS     
gateway      1          5          deny     deny     0         
firewall     1          0          deny     stateful allow 0         
router       0          2          allow    deny     0         
nat          1          1          allow    allow    0         
overlay      1          1          deny     deny     0         
```
Outgoing packets are processed by the layers going top-to-bottom and incoming packets go through the reverse. That bottom layer, `overlay`, is where the GENEVE encap/decap happens.

**Egress:**

Outgoing packets are wrapped with outer ETH/IPv6/UDP/GENEVE headers. The outer IPv6 source is the sled's underlay address & the destination is the boundary services address `fd00:99::1`. The UDP destination port is always 6081 (the IANA assigned port for Geneve) and the source port we just always set to 7777 (the Zezeski special 😅️). We also stash a VPC-unique VNI into the Geneve header. The outer ethernet frame addresses get populated based on the entries in the illumos routing table. That's what the entries for `fd00:99::1` on both `cxgeb0` & `cxgeb1` are for. AFAIK, we should be choosing one probabilistically using DDM (Delay Driven Multipath). Either way, we'll set the destination address to a gateway on one of the chelsios and use its MAC as the source address.

Once the packet gets to the switch, it recognizes it as a boundary services destined geneve encapsulated payload and will strip the outer headers. From there, it'll use the uplink configuration like routing and available links to send it out to the customer network as appropriate.

**Ingress:**

For incoming packets from the customer network, they'll hit the switch which will check it's NAT table to determine which sled to forward the packet to:
```
 pfexec zlogin oxz_switch /opt/oxide/dendrite/bin/swadm nat list
External IP   Port low  Port high  Internal IP            Inner MAC          VNI
172.17.16.11  0         65535      fd00:1122:3344:101::1  a8:40:25:ff:ef:ce  100
172.17.16.12  0         65535      fd00:1122:3344:101::1  a8:40:25:ff:db:ff  100
172.17.16.13  0         65535      fd00:1122:3344:101::1  a8:40:25:ff:b2:cc  100
172.17.16.14  0         65535      fd00:1122:3344:101::1  a8:40:25:ff:be:a1  100
172.17.16.15  0         16383      fd00:1122:3344:101::1  a8:40:25:ff:e2:15  100
```
If there's an entry that matches the destination IP and port, it'll similarly add outer ETH/IPv6/UDP/GENEVE headers. The outer IPv6 destination will be the "Internal IP" from the NAT entry which represents a given sled's underlay address (the sled hosting that instance/service). The outer ETH destination will be its corresponding value from the switch's NDP table. Similarly to the egress path, the VNI will also be included in the Geneve header. But that's not enough to identify which specific OPTE port on the sled should receive the packet, so it'll also update the _inner_ ETH destination to the "Inner MAC" from the NAT entry. We can see the MAC on the OPTE ports should match the NAT entries:
```
 pfexec opteadm list-ports                                                                                                                    
LINK                             MAC ADDRESS              IPv4 ADDRESS     EXTERNAL IPv4    IPv6 ADDRESS                             EXTERNAL IPv6                            STATE   
opte0                            A8:40:25:FF:E2:15        172.30.3.5       None             None                                     None                                     running 
opte1                            A8:40:25:FF:EF:CE        172.30.1.5       172.17.16.11     None                                     None                                     running 
opte5                            A8:40:25:FF:BE:A1        172.30.2.7       172.17.16.14     None                                     None                                     running 
opte6                            A8:40:25:FF:B2:CC        172.30.2.6       172.17.16.13     None                                     None                                     running 
opte7                            A8:40:25:FF:DB:FF        172.30.2.5       172.17.16.12     None                                     None                                     running 
```
At this point we'll hit the `overlay` in OPTE as an inbound packet this time where those outer headers will be stripped and further NAT, firewall rules get applied before being delivered to the instance/service's vnic.

Reply:
@luqmana This is very helpful though I didn't follow all of it.  I'm not sure how best to incorporate it.  Do you want to make a PR on this PR, or just push to this PR, or add this in a follow-up PR?

Reply:
> @luqmana This is very helpful though I didn't follow all of it. I'm not sure how best to incorporate it. Do you want to make a PR on this PR, or just push to this PR, or add this in a follow-up PR?

Also, please call out any bits that weren't clear enough!